### PR TITLE
feat(rpc): add compression to JSON-RPC

### LINKED ISSUE #9652: Enable GZIP compression for JSON-RPC calls
### Describe the feature

Introducing gzip compression for `JSON-RPC` calls can significantly `reduce` the payload size of the requests and responses, leading to faster communication and lower bandwidth usage. This enhancement can be particularly beneficial for applications making frequent or large JSON-RPC calls to the Ethereum network.

```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by TheDhejavu**
## Description
Resolves: #9652

This adds a compression layer that wraps [`tower_http::compression::CompressionLayer`](https://github.com/tower-rs/tower-http/blob/main/tower-http/src/compression/layer.rs) and adapts it to work with jsonrpsee's HTTP types. The layer automatically compresses responses based on the client's `Accept-Encoding` header while preserving the middleware chain's expected body type.

Supports:
  - `gzip`
  - `deflate`
  - `brotli`
  - `zstd`

### Demo
Screenshot 2024-11-11 at 11 35 25 AM

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, but I think hyper doesn't apply any changes to the body automatically so we'd need to do that

this seems relevant:

https://github.com/paritytech/jsonrpsee/issues/1133

**Comment by TheDhejavu** [comment]
> cool, but I think hyper doesn't apply any changes to the body automatically so we'd need to do that
> 
> this seems relevant:
> 
> [paritytech/jsonrpsee#1133](https://github.com/paritytech/jsonrpsee/issues/1133)

Thanks!!

figured out that reth has `tower-http` as a dependency and compression middleware is implemented by default which works out-of-the-box with `jsonrpee` with their recent release [here](https://github.com/paritytech/jsonrpsee/releases/tag/v0.24.7). 

in the case of this: https://github.com/paritytech/jsonrpsee/issues/1133#issuecomment-1562534957, the implementation is quite similar to what i did here but i will need to modify the body in order to compress it.

this will be the only change i will probably need to make if i use tower compression layer implementation.
```rust
use tower_http::compression::CompressionLayer;

impl<RpcMiddleware> RpcServerConfig<RpcMiddleware> {
    //.....other middlewares

    /// Returns a [`CompressionLayer`] that adds compression support (gzip, deflate, brotli, zstd)
    /// based on the client's `Accept-Encoding` header
    const fn maybe_compression_layer() -> Option<CompressionLayer> {
          Some(CompressionLayer::new())
    }
}
```

wdyt ?

**Comment by mattsse** [comment]
cool, sgtm!

**Comment by TheDhejavu** [comment]
cc @mattsse .

File: crates/rpc/rpc-layer/src/compression_layer.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    fn layer(&self, inner: S) -> Self::Service {
+        CompressionService { compression: self.inner_layer.layer(inner) }
+    }
+}
+
+#[allow(missing_debug_implementations)]
+#[derive(Clone)]
+/// Service that performs response compression.
+///
+/// Created by [`CompressionLayer`].
```
Comment (mattsse) [inline crates/rpc/rpc-layer/src/compression_layer.rs:48]: same

**Review by mattsse — CHANGES_REQUESTED** [review]
great!

only have pedantic style nits

File: Cargo.toml [hunk 2]
```diff
@@ -553,7 +553,8 @@ hyper-util = "0.1.5"
 pin-project = "1.0.12"
 reqwest = { version = "0.12", default-features = false }
 tower = "0.4"
-tower-http = "0.5"
+tower-http = { version = "0.6", features = ["full"] }
```
[the remark this answers is not in the document]
Comment (TheDhejavu) [reply Cargo.toml]: removed, but i left the upgrade.

**Review by mattsse — APPROVED** [review]

**Pull request merged**